sdk/scheduler/src/main/java/com/mesosphere/sdk/offer/DefaultOfferRequirementProvider.java

```diff
@@ -318,11 +315,56 @@ private static void validateTaskRequirements(List<TaskRequirement> taskRequireme
             }
         }
 
-        return resources;
+        return coalescePorts(resources);
+    }
+
+    private static List<Protos.Resource> coalescePorts(Collection<Protos.Resource> resources) {
+        // Within the SDK, each port is handled as its own resource, since they can have extra meta-data attached, but
+        // we can't have multiple "ports" resources on a task info, so we combine them here. Since ports are also added
+        // back onto TaskInfos during the evaluation stage (since they may be dynamic) we actually just clear the ranges
+        // from that resource here to make the bookkeeping easier.
+        Protos.Resource portsResource = null;
+        List<Protos.Resource> coalescedResources = new ArrayList<>();
+
+        for (Protos.Resource r : resources) {
+            if (r.getName().equals("ports") && portsResource == null) {
```

Comment:
There's a `Constants.java` file.  Might be good to make a constant for "ports" there.  If not there, somewhere.